### PR TITLE
Use the ModuleInterface interface instead of class Module

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Because of interface 🤙.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | No need QA
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

## BC Breaks

 - Add `ModuleInterface::getInstance()`
 - Add `ModuleInterface::hasValidInstance()`
 - Add `ModuleInterface::get(string $name)`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26571)

### CODE REVIEW

**Review by matks — APPROVED** [review]
Interface is always better than implementation.

But this PR contains BC breaks, if I am not wrong ? 🤔  or do you consider than replacing a type-hint of class A by interface B (if A implements B) is OK?

**Comment by PierreRambaud** [comment]
> Interface is always better than implementation.
> 
> But this PR contains BC breaks, if I am not wrong ? thinking or do you consider than replacing a type-hint of class A by interface B (if A implements B) is OK?

It's not a BC break as using the `Module` class is still working and doesn't break anything.
Also, even if you extend the class, it's still working.
http://sandbox.onlinephpfunctions.com/code/5f3a3eee170b03f21c0aed1edc1034f820cdaecd

File: src/Core/Addon/Module/ModuleInterface.php [hunk 1]
```diff
@@ -30,6 +30,12 @@
 
 interface ModuleInterface extends AddonInterface
 {
+    public function getInstance();
+
+    public function hasValidInstance();
+
+    public function get(string $name);
+
```
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleInterface.php:38]: Please document the BC Break

Comment (Progi1984) [reply src/Core/Addon/Module/ModuleInterface.php:38]: Isn't it possible to typehint the return of methods ?

**Review by Progi1984 — COMMENTED** [review]
@PierreRambaud Please document BC Breaks (all signatures changes)

File: src/Core/Addon/Module/ModuleInterface.php (hunk 1 above)
Comment (PierreRambaud) [reply src/Core/Addon/Module/ModuleInterface.php:38]: The `Core/Addon` namespace will be completely removed in the 8.0

Comment (Progi1984) [reply src/Core/Addon/Module/ModuleInterface.php:38]: It's currently a BC Break. The removal has not been done.

**Review by Progi1984 — APPROVED** [review]

**Pull request merged**